imaginairy/img_processors/control_modes.py

```diff
@@ -232,6 +232,51 @@ def noop(img):
     return (img + 1.0) / 2.0
 
 
+def otsu_binarize(img: "Tensor") -> "Tensor":
+    """
+    Use's Otsu's method to select the threshold to binarize the image.
+
+    Chose not to use a 3rd party library for otsu's method because it gave inconsistent results depending
+    on the device it was used on. Also, this implementation gives better results than the 3rd party libraries.
+    """
+    import numpy as np
```

Comment:
see if this torch only version works:
```python
def otsu_binarize(img: "torch.Tensor") -> "torch.Tensor":
    """
    Use's Otsu's method to select the threshold to binarize the image.
    This version uses PyTorch operations instead of NumPy.
    """
    import torch

    from imaginairy.utils import get_device

    img = img.to(get_device())

    if img.dim() != 4:
        raise ValueError("Input should be a 4d tensor")
    if img.size(1) != 3:
        raise ValueError("Input should have 3 channels")

    if not torch.all((img >= -1) & (img <= 1)):
        raise ValueError("All tensor values must be between -1 and 1")

    normalized = (img + 1) / 2
    grayscale = normalized.mean(dim=1, keepdim=True)

    flat_gray = grayscale.flatten()

    # Histogram and bin centers
    hist = torch.histc(flat_gray, bins=256, min=0, max=1)
    bin_edges = torch.linspace(0, 1, steps=257)
    bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2

    first_weight = torch.cumsum(hist, dim=0)
    second_weight = torch.cumsum(hist.flip(dims=[0]), dim=0).flip(dims=[0])
    first_mean = torch.cumsum(hist * bin_centers) / first_weight
    second_mean = (torch.cumsum((hist * bin_centers).flip(dims=[0])) / second_weight.flip(dims=[0])).flip(dims=[0])

    inter_class_variance = first_weight[:-1] * second_weight[1:] * (first_mean[:-1] - second_mean[1:]) ** 2
    optimal_threshold = bin_centers[:-1][torch.argmax(inter_class_variance)]

    binary = (grayscale > optimal_threshold).float()
    return binary.repeat(1, 3, 1, 1)
```

Reply:
see if cv2.adaptiveThreshold works better

Reply:
I'm using cv2.adpativeThreshold now but according to chatgpt it needs numpy input.